src/backend/env/post.rs

```diff
@@ -527,30 +554,81 @@ impl Post {
     }
 
     pub fn count(state: &State) -> usize {
-        state.posts.len()
+        state.posts.len() + state.memory.posts.len()
     }
 
+    // Get the post from the heap if available, or load from the stable memory into the cache and
+    // return the refernce to it
     pub fn get<'a>(state: &'a State, post_id: &PostId) -> Option<&'a Post> {
-        state.posts.get(post_id)
+        state.posts.get(post_id).or_else(|| {
+            let boxed = cache().get(post_id).or_else(|| {
+                state.memory.posts.get(post_id).and_then(|mut post: Post| {
+                    let cache = cache();
+                    post.archived = true;
+                    cache.insert(*post_id, Box::new(post));
+                    cache.get(post_id)
+                })
+            });
+            boxed.map(|ptr| &**ptr)
+        })
     }
 
+    // Takes the post from cold or hot memory
+    fn take(state: &mut State, post_id: &PostId) -> Result<Post, String> {
+        cache().remove(post_id);
+        state
+            .posts
+            .remove(post_id)
+            .ok_or("no post found".to_string())
+            .or_else(|_| state.memory.posts.remove(post_id))
+    }
+
+    // Takes the post from hot or cold memory, mutates and inserts into the hot memory
     pub fn mutate<T, F>(state: &mut State, post_id: &PostId, f: F) -> Result<T, String>
     where
         F: FnOnce(&mut Post) -> Result<T, String>,
     {
-        let mut post = state.posts.remove(post_id).ok_or("no post found")?;
+        let mut post = Post::take(state, post_id)?;
         let result = f(&mut post);
-        if state.posts.insert(*post_id, post).is_some() {
-            panic!("no post should exist")
-        }
+        Post::save(state, post);
         result
     }
 
     fn save(state: &mut State, post: Post) {
-        state.posts.insert(post.id, post);
+        if state.posts.insert(post.id, post).is_some() {
+            panic!("no post should exist")
+        }
     }
 }
 
+// Moves a configured number of posts from hot to cold memory.
+pub fn archive_cold_posts(state: &mut State, max_posts_in_heap: usize) -> Result<(), String> {
+    let mut posts: Vec<(u64, PostId)> = state
+        .posts
+        .values()
+        .map(|post| (post.timestamp(), post.id))
+        .collect();
+    posts.sort_unstable_by(|(t1, _), (t2, _)| t1.cmp(t2));
```

Comment:
I'd just get the timestamp in the sorting function
```suggestion
        .collect();
    posts.sort_unstable_by(|(p1, _), (p2, _)| p1.timestamp().cmp(p2.timestamp()));
```

Reply:
Good idea, thanks Christian!